README.rst
Scan PDF - Easy scans in Linux with a document scanner like the Fujitsu ScanSnap
################################################################################

.. image:: http://badge.fury.io/py/scanpdf.png
    :target: http://badge.fury.io/py/scanpdf

.. image:: http://pypip.in/d/scanpdf/badge.png
    :target: https://crate.io/packages/scanpdf?version=latest


If you're looking for a simple way to use a multi-page scanner and get your
document into a PDF in Linux without any proprietary or commercial software,
then ScanPDF might be the solution.  I wrote it to quickly take the Linux SANE
scanner system output image files, and process them into usable PDFs.  By
usable, I mean PDFs that maintain their original scanned resolution, omit blank
pages (if you're scanning in duplex mode, for example), preserve color unless
the original is greyscale/black and white, in which case they are intelligently
down-converted to B/W PDFs to save space.

* Free and open-source software: ASL2 license
* Documentation: http://virantha.github.io/scanpdf/html
* Source: https://github.com/virantha/scanpdf

Features
--------
* Uses SANE/scanadf to automatically scan to multi-page compressed PDFs
* `Integrates with ScanBd <http://virantha.github.io/scanpdf/html>`_ to respond to hardware button presses
* Automatically removes blank pages.
* Scans in color, and automatically down-converts into 1-bit B/W image for text/greyscale images
* (optionally) Auto-crops to the proper page size.
* (optionally) applyies unpaper formatting to finished images
* (optionally) applies pdf-sandwich text-recognition to finished pdf

Usage:
------
The simplest way to use this is:

::

    scanpdf scan pdf <pdffile>

    Or alternatively

    scanpdf scan pdf (To bring up a file-save dialog to direct the finished pdf file.)

This will first perform the scan, and then the conversion to PDF.  If you want
to split up the scan and the PDF conversion into two separate invocations (for
reasons clarified below), then you can do:

::

    scanpdf --tmpdir=tmp scan
    scanpdf --tmpdir=tmp pdf <pdffile>
  
One reason for the separation might be if you want to keep scanning documents
(very quick) while the post-processing (slower) for the PDF conversion is
taking place in the background.   For instance, if you're using the hardware
button on the scanner to initiate scans (as detailed in this_ document), then
you want to return immediately after the scan instead of waiting for the full
conversion to PDF has taken place.

.. _this: http://virantha.com/2014/03/17/one-touch-scanning-with-fujitsu-scansnap-in-linux/

You can optionally use the following switches to control if you're putting pages face up or face down in the auto
document feeder, if you want to skip the blank page processing, adjust the blank page detection threshold, or add 
additional post-processing using unpaper_:

.. _unpaper: http://unpaper.berlios.de

::

        --dpi=<dpi>                 DPI to scan in [default: 300]
        --device=<device>           Scanning device (sub '%' for spaces)
        --crop                      Run ImageMagick cropping routine
        --tmpdir=<dir>              Temporary directory
        --keep-tmpdir               Whether to keep the tmp dir after scanning or not [default: False]
        --face-up=<true/false>      Face-up scanning [default: True]
        --keep-blanks               Don't check for and remove blank pages
        --blank-threshold=<ths>     Percentage of white to be marked as blank [default: 0.97]
        --post-process              Process finished images with unpaper
        --text-recognize            Run pdfsandwich for text recognition

Right now, I'm assuming this is getting called via ScanBD, so I don't have the option to manually specify the
scanner.  If you really want to use this standalone, for now, either set the --device option or just set the
         ``SCANBD_DEVICE`` environment variable to your scanner device name before running this script.


Installation
------------
::

    $ pip install scanpdf

Requires ImageMagick and SANE to be installed, for the command line tools:

* ``convert``
* ``identify``
* ``ps2pdf``
* ``scanadf``
* ``unpaper``
* ``pdfsandwich``

Also requires epstopdf.

Disclaimer
----------
The software is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.


CHANGES.rst
=======  ========   ======
Version  Date       Changes
-------  --------   ------

v0.4.0   9/22/16    Add deskew, text-recognition, gui-dialog for save, multithreading
v0.3.0   8/25/14    Allow arbitrary page sizes and auto-crops
v0.1.0   1/1/14     First release
=======  ========   ======


scanpdf/test_gui.py
import shutil
import tkFileDialog
import os

OUTPUT_DIRECTORY = "/home/dale/Temp"
FILENAME = "image_test.pdf"

def file_save(source_file):
    f = tkFileDialog.asksaveasfile(mode='w', defaultextension=".pdf")
    if f is None: # asksaveasfile return `None` if dialog closed with "cancel".
        return
    target_path = f.name
    f.close()
    os.remove(target_path)
    shutil.move(source_file, target_path)



def main():
    file_save(os.path.join(OUTPUT_DIRECTORY, FILENAME))

if __name__ == '__main__':
    main()


test/checkscanpdf.py
#!/usr/bin/env python2.7

# Basic Test for scanpdf.py

import os
from scanpdf import scanpdf

OUTPUT_DIRECTORY = "~/Temp"
FILENAME = "image_test.pdf"


def ensure_dir(f):
    d = os.path.dirname(f)
    if not os.path.exists(d):
        os.makedirs(d)


def main():
    ensure_dir(OUTPUT_DIRECTORY)
    full_path = os.path.join(OUTPUT_DIRECTORY, FILENAME)
    if os.path.exists(full_path):
        os.remove(full_path)
    os.environ["SCANBD_DEVICE"] = 'net:localhost:fujitsu:ScanSnap S1500:1448'
    args = {'scan': True, 'pdf': True, '<pdffile>': full_path,
            '--verbose': True, '--debug': True, '--device': "net:localhost:fujitsu:ScanSnap S1500:1448",
            '--dpi': 300, '--crop': False, '--tmpdir': False,
            '--keep-blanks': False, '--blank-threshold': 0.80, '--text-recognize': False,
            '--face-up': False, '--keep-tmpdir': False, '--post-process': False}
    script = scanpdf.ScanPdf()
    print args
    script.go(args)


if __name__ == '__main__':
    main()


test/test_scanpdf.py
import scanpdf.ScanPdf as P
import pytest
import os
import logging

import smtplib
from mock import Mock
from mock import patch, call
from mock import MagicMock
from mock import PropertyMock


class Testscanpdf:

    def setup(self):
        self.p = P.ScanPdf()


fabfile.py
from fabric.api import *
import os, sys
 
project_dir = os.path.join(os.path.dirname(sys.argv[0]))

def build_windows_dist():
    if os.name == 'nt':
        # Call the pyinstaller
        local("python ../pyinstaller/pyinstaller.py scanpdf_windows.spec --onefile")


def run_tests():
    test_dir = "test"
    with lcd(test_dir):
        # Regenerate the test script
        local("py.test --genscript=runtests.py")
        t = local("py.test --cov-config .coveragerc --cov=scanpdf --cov-report=term --cov-report=html", capture=False)

        with open("test/COVERAGE.rst", "w") as f:
            f.write(t)


def push_docs():
    """ Build the sphinx docs from develop
        And push it to gh-pages
    """
    githubpages = "/Users/virantha/dev/githubdocs/scanpdf"
    # Convert markdown readme to rst
    #local("pandoc README.md -f markdown -t rst -o README.rst")
    with lcd(githubpages):
        local("git checkout gh-pages")
        local("git pull origin gh-pages")
    with lcd("docs"):
        print("Running sphinx in docs/ and building to ~/dev/githubpages/scanpdf")
        local("make clean")
        local("make html")
        #local("cp -R ../test/htmlcov %s/html/testing" % githubpages)
    with lcd(githubpages):
        local("git add .")
        local('git commit -am "doc update"')
        local('git push origin gh-pages')


html/searchindex.js
Search.setIndex({envversion:42,terms:{oper:2,all:2,identifi:5,help:2,just:[2,5],whoever:2,text:5,offici:2,scanpdf:[2,5,1],none:4,press:5,through:2,file:[2,5],threshold:5,follow:5,yet:4,virtualenvwrapp:2,web:2,scansnap:0,system:[2,5],easi:0,how:2,duplex:5,explain:2,solut:5,send:2,articl:2,blog:2,add:[2,5,6],version:2,black:5,scanadf:5,skip:5,save:5,local:2,easier:2,split:5,your_name_her:2,them:[2,5],sourc:5,everi:2,which:5,thei:[2,5],python:2,auto:5,express:5,pypi:[2,7],kind:5,initi:5,repo:2,greatli:2,mkvirtualenv:2,background:5,pull_request:2,fab:2,bit:[2,5],now:[2,5],requir:5,like:0,down:5,term:6,name:[2,5],anyth:2,pdffile:5,troubleshoot:2,list:[2,0],separ:5,whether:2,sane:5,mode:5,quick:5,output:5,updat:2,button:5,wrote:5,mean:5,pleas:[2,5],impli:5,set:[2,5],slower:5,intellig:5,manual:5,linux:0,crop:5,realli:5,virtualenv:2,sure:2,greyscal:5,pass:2,proper:5,imagemagick:5,todo:0,best:2,gmail:4,even:2,index:0,detect:5,space:5,preserv:5,open:[2,5],http:[2,5],"while":5,content:0,adjust:5,simplest:5,hardwar:5,"new":2,fujitsu:0,reproduc:2,airfram:[2,3],integr:5,clarifi:5,full:5,run:5,appreci:2,gener:6,usag:0,given:2,here:2,disclaim:0,standalon:5,step:2,scanbd:5,free:5,put:[2,5],org:2,post:[2,5],addit:5,convert:5,deskew:5,face:5,could:2,omit:5,alon:6,reason:5,page:[5,0],instanc:5,place:5,dpi:5,commerci:5,fals:5,commit:2,chang:2,narrow:2,com:[2,5,4],support:2,first:[5,7,4],origin:[2,5],softwar:5,via:5,volunt:2,virantha:[2,5,4],color:5,search:0,modul:0,automat:5,feeder:5,warranti:5,releas:7,right:5,done:2,littl:2,blank:5,"import":3,should:2,assum:[2,5],your:[2,5],size:5,take:5,tox:2,git:2,fabric:2,wai:[2,5],script:5,licens:5,devic:5,two:5,coverag:0,mark:5,websit:2,readi:2,mani:2,call:5,includ:2,taken:5,scope:2,white:5,wait:5,instal:[2,0],more:[2,6],fork:2,setup:2,option:5,travi:2,imag:5,resolut:5,about:2,tmpdir:5,thi:[2,5],part:2,percentag:5,might:[2,5],checkout:2,rst:2,line:[5,1],stand:6,"true":5,convers:5,"case":5,multi:5,immedi:5,look:[2,5],possibl:2,"default":5,bugfix:2,remov:5,work:2,unpap:5,exampl:5,project:[2,3],below:5,respond:5,can:[2,5],doc:2,meet:2,invoc:5,quickli:5,control:5,"function":2,would:2,scanner:0,creat:2,scan:2,process:5,readm:2,ani:[2,5],rememb:2,tag:2,proprietari:5,have:[2,5],command:[5,1],pip:[2,5,1],"return":5,keep:[2,5],check:[2,5],alwai:2,asl2:5,welcom:2,want:[2,5],perform:5,make:[2,6],when:2,detail:[2,5],also:5,html:5,ps2pdf:5,branch:2,test:2,instead:5,you:[2,5],tmp:5,simpl:5,tool:5,basi:5,either:5,variabl:5,copi:2,unless:5,distribut:5,clone:2,after:5,compress:5,docstr:[2,6],usabl:5,driven:2,befor:[2,5],run_test:2,scanbd_devic:5,pdf:2,veri:5,ekanayak:[2,4],why:4,github:[2,5],don:5,specifi:5,condit:5,descript:2,without:5,issu:2,"switch":5,maintain:5,environ:5,epstopdf:5,clean:5,push:2,propos:2},objtypes:{},objnames:{},filenames:["index","installation","contributing","usage","authors","readme","todo","history"],titles:["Scan PDF API Reference (version 0.3.1)","Installation","Contributing","Usage","Credits","Scan PDF - Easy scans in Linux with a document scanner like the Fujitsu ScanSnap","Todo list","History"],objects:{},titleterms:{featur:[2,5],feedback:2,scan:[5,0],guidelin:2,linux:5,request:2,indic:0,api:0,tabl:0,contributor:4,instal:[5,1],todo:6,scanner:5,develop:4,lead:4,fix:2,easi:5,tip:2,submit:2,write:2,start:2,version:0,test:0,document:[2,5],type:2,refer:0,contribut:2,get:2,usag:[3,5],disclaim:5,report:2,scansnap:5,bug:2,pull:2,like:5,list:6,histori:7,credit:4,fujitsu:5,pdf:[5,0],implement:2}})

scanpdf/scanpdf.py
#!/usr/bin/env python2.7
# Copyright 2014 Virantha Ekanayake All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Scan to PDF.

Usage:
    scanpdf [options] scan 
    scanpdf [options] pdf <pdffile> 
    scanpdf [options] scan pdf <pdffile>
    scanpdf [options] pdf
    scanpdf [options] scan pdf


Options:
    -v --verbose                Verbose logging
    -d --debug                  Debug logging
    --dpi=<dpi>                 DPI to scan in [default: 300]
    --device=<device>           Scanning device (sub '%' for spaces)
    --crop                      Run ImageMagick cropping routine
    --tmpdir=<dir>              Temporary directory 
    --keep-tmpdir               Whether to keep the tmp dir after scanning or not [default: False]
    --face-up=<true/false>      Face-up scanning [default: True]
    --keep-blanks               Don't check for and remove blank pages
    --blank-threshold=<ths>     Percentage of white to be marked as blank [default: 0.97]
    --post-process              Process finished images with unpaper
    --text-recognize            Run pdfsandwich for text recognition
    
"""

import glob
import logging
import multiprocessing
import os
import re
import shutil
import subprocess
import sys
import time
import wx
from multiprocessing.dummy import Pool as Threadpool

import docopt

from version import __version__

date_format = '%m/%d/%Y %H:%M:%S'


class ProcessPage:
    """Process page applies methods to a single PPM Image
    Attributes:
        page (str): Filename of image
        scanpdf (ScanPdf): ScanPDF object which instantiated the ProcessPage
    """
    page = None
    scanpdf = None

    def __init__(self, page, scanpdf):
        """
        """
        self.page = page
        self.scanpdf = scanpdf
        os.chdir(self.scanpdf.tmp_dir)

    def process(self):
        """
        Apply all processing in accordance with selcted options
        :return: None
        """
        self.run_deskew()
        if self.scanpdf.crop:
            self.run_crop()
        self.convert_to_bw()
        if not self.scanpdf.keep_blanks:
            self.remove_blank()
        if self.page is not None and self.scanpdf.post_process:
            self.run_postprocess()

    def run_deskew(self):
        """
        Deskew image using Marke Mauder Deskew
        https://bitbucket.org/galfar/app-deskew
        :return: None
        """
        deskew = os.path.dirname(os.path.realpath(__file__)) + os.path.sep + 'deskew64'
        logging.info("Deskewing: " + os.path.basename(self.page))
        ppm_page = '%s.ppm' % self.page
        c = [deskew, ' %s ' % self.page, '-o', ppm_page]
        result = self.scanpdf.cmd(c)
        logging.debug("deskew result: " + result)
        os.remove(self.page)
        self.page = ppm_page

    def run_crop(self):
        """
        Apply standard cropping routine from ImageMagick
        :return:
        """
        logging.info("Cropping: " + os.path.basename(self.page))
        crop_page = '%s.crop' % self.page
        c = ['convert', '-fuzz 20%', '-trim', self.page, crop_page]
        self.scanpdf.cmd(c)
        os.remove(self.page)
        self.page = crop_page

    def run_postprocess(self):
        logging.info("Post-processing with unpaper: " + os.path.basename(self.page))
        shutil.move(self.page, '%s.ppm' % self.page)
        self.page = '%s.ppm' % self.page
        processed_page = '%s_unpaper' % self.page
        c = ['unpaper', self.page, processed_page]
        self.scanpdf.cmd(c)
        os.remove(self.page)
        self.page = processed_page

    def convert_to_bw(self):
        """
        Check if color exists, convert to black-white if not
        :return: None
        """
        logging.info("Checking if: " + os.path.basename(self.page) + " is bw...")
        if not self._is_color():
            self._page_to_bw()

    def _page_to_bw(self):
        """
        Apply ImageMagick Black-White Conversion
        :return: None
        """
        bw_page = "%s_bw" % self.page
        c = "convert %s +dither -colors 2 -colorspace gray -normalize %s_bw" % (self.page, self.page)
        self.scanpdf.cmd(c)
        # Remove the old file
        os.remove(self.page)
        self.page = bw_page

    @staticmethod
    def run(args):
        """
        Static run method for ProcessPage
        :param args: page (string pagename), ScanPdf tuple
        :return: name of final processed page
        """
        process_page = ProcessPage(args[0], args[1])
        process_page.process()
        return process_page.page

    def remove_blank(self):
        """
        Check if page is blank: if so, remove
        :return: None
        """
        logging.info("Checking if: " + os.path.basename(self.page) + " is blank")
        if self.is_blank():
            os.remove(self.page)
            self.page = None

    def is_blank(self):
        """
        Check if page blank by comparing stdev from ImageMagick 'Identify'
        to threshold (1 - blank_threshold)
        :return: true if image is blank
        """
        if not os.path.exists(self.page):
            return True
        c = 'identify -verbose %s' % self.page
        result = self.scanpdf.cmd(c)
        m_std_dev = re.compile("""\s*standard deviation:\s*\d+\.\d+\s*\((?P<percent>\d+\.\d+)\).*""")
        for line in result.splitlines():
            match = m_std_dev.search(line)
            if match:
                stdev = float(match.group('percent'))
                logging.info(os.path.basename(self.page) + " std. dev: " + "{0:.4f}".format(stdev))
                if stdev > 1. - self.scanpdf.blank_threshold:
                    return False
        return True

    def _is_color(self):
        """
        Run the following command from ImageMagick:
        convert holi.pdf -colors 8 -depth 8 -format %c histogram:info:-
        This outputs something like the following:
        10831: ( 24, 26, 26,255) #181A1A srgba(24,26,26,1)
        4836: ( 55, 87, 79,255) #37574F srgba(55,87,79,1)
        6564: ( 77,138,121,255) #4D8A79 srgba(77,138,121,1)
        4997: ( 86, 96, 93,255) #56605D srgba(86,96,93,1)
        7005: ( 92,153,139,255) #5C998B srgba(92,153,139,1)
        2479: (143,118,123,255) #8F767B srgba(143,118,123,1)
        8870: (169,176,170,255) #A9B0AA srgba(169,176,170,1)
        442906: (254,254,254,255) #FEFEFE srgba(254,254,254,1)
        1053: (  0,  0,  0,255) #000000 black
        484081: (255,255,255,255) #FFFFFF white
        :return: true if image color exists
        """
        c = "convert %s -colors 8 -depth 8 -format %%c histogram:info:-" % self.page
        out = self.scanpdf.cmd(c)
        m_line = re.compile(r"""\s*(?P<count>\d+):\s*\(\s*(?P<R>\d+),\s*(?P<G>\d+),\s*(?P<B>\d+).+""")
        colors = []
        for line in out.splitlines():
            match_line = m_line.search(line)
            if match_line:
                logging.debug("Found RGB values")
                color = [int(x) for x in (match_line.group('count'),
                                          match_line.group('R'),
                                          match_line.group('G'),
                                          match_line.group('B'),
                                          )
                         ]
                colors.append(color)
        # sort
        colors.sort(reverse=True, key=lambda y: y[0])
        logging.debug(colors)
        is_color = False
        logging.debug(colors)
        for color in colors:
            # Calculate the mean differences between the RGB components
            # Shades of grey will be very close to zero in this metric...
            diff = float(sum([abs(color[2] - color[1]),
                              abs(color[3] - color[1]),
                              abs(color[3] - color[2]),
                              ])) / 3
            if diff > 20:
                is_color = True
                logging.debug("Found color")
        return is_color


class ScanPdf(object):
    """
    Controls Scan and PDF functions
    Attributes:
        pages (list): list of page filenames
        cwd (str): Current Working Directory
        tmp_dir(str): name of temporary directory for original and processed images
        args (dict): command line arguments received from docopt
        device (str): scanning device name recongized by scanadf
        pdf_filename (str): path/name for pdf output
        dpi (int): resolution for scanned image (dots per inch)
        keep_blanks (bool): Whether to keep blank images
        crop (bool): Whether to crop images
        post-process (bool): Whether to apply unpaper post-processing
        text_recognize (bool): Whether to apply pdf-sandwich post processing
    """
    pages = None
    cwd = None
    tmp_dir = None
    args = None
    device = None
    pdf_filename = None
    dpi = None
    keep_blanks = None
    blank_threshold = None
    crop = None
    post_process = None
    text_recognize = None

    def __init__(self):
        """ 
        """
        self.config = None
        self.cwd = os.getcwd()

    def cmd(self, cmd_list):
        """
        Run command in operating system
        :param cmd_list: list of commands
        :return: os outpout of command
        """
        if isinstance(cmd_list, list):
            cmd_list = ' '.join(cmd_list)
        logging.debug("Running cmd: %s" % cmd_list)
        try:
            out = subprocess.check_output(cmd_list, stderr=subprocess.STDOUT, shell=True)
            logging.debug(out)
            return out
        except subprocess.CalledProcessError as e:
            print e.output
            self._error("Could not run command %s" % cmd_list)

    def run_scan(self):
        """
        Scan documents from device
        :return: None
        """
        if self.device is None:
            self._error("Scanning device is undefined!")
        self.cmd('logger -t "scanbd: " "Begin of scan "')
        c = ['scanadf',
             '-d "%s"' % self.device,
             '--source "ADF Duplex"',
             '--mode Color',
             '--resolution %sdpi' % self.dpi,
             '-o %s/page_%%04d' % self.tmp_dir,
             '-y 279.364',
             '--page-height 279.364',
             ]
        self.cmd(c)
        self.cmd('logger -t "scanbd: " "End of scan "')
        file_count = len([name for name in os.listdir(self.tmp_dir) if
                          os.path.isfile(os.path.join(self.tmp_dir, name))])
        logging.info('Receved {0:d} files in {1:}...'.format(file_count, self.tmp_dir))

    @staticmethod
    def _error(msg):
        """
        print error and exit
        :param msg: error message
        :return: None
        """
        print("ERROR: %s" % msg)
        sys.exit(-1)

    @staticmethod
    def _atoi(text):
        return int(text) if text.isdigit() else text

    def _natural_keys(self, text):
        """
         alist.sort(key=natural_keys) sorts in human order
         http://nedbatchelder.com/blog/200712/human_sorting.html
         (See Toothy's implementation in the comments)
        :param text:
        :return: sorted
        """
        return [self._atoi(c) for c in re.split('(\d+)', text)]

    def get_pages(self):
        cwd = os.getcwd()
        os.chdir(self.tmp_dir)
        pages = glob.glob('page_*')
        pages.sort(key=self._natural_keys)
        os.chdir(cwd)
        return pages

    def reorder_face_up(self):
        assert len(self.pages) % 2 == 0, "Why is page count not even for duplexing??"
        logging.info("Reordering pages")
        # for i in range(0,len(pages),2):
        # pages[i], pages[i+1] = pages[i+1], pages[i]
        self.pages.reverse()

        # OLD CODE - doesn't work for color images
        # c = 'convert %s -shave 1%%x1%%  -format "%%[fx:mean]" info:' % filename
        # result = self.cmd(c)
        # if float(result.strip()) > self.blank_threshold:
        #     return True
        # else:
        #     return False

    def run_text_recognize(self, pdf_file):
        """
        run text recognition
        :param pdf_file: pdf filename (in temporary directory)
        :return:
        """
        c = ['pdfsandwich', '-coo', '\"-deskew 40%\"', pdf_file]
        self.cmd(c)
        filename, file_extension = os.path.splitext(pdf_file)
        ocr_file = filename + "_ocr" + file_extension
        os.remove(pdf_file)
        shutil.move(ocr_file, pdf_file)

    @staticmethod
    def save_file(source_file):
        """
        Run dialog to save file to user-defined name/folder
        :param source_file: pdf file in temporary directory
        :return: None
        """
        path = None
        _ = wx.App(redirect=True)
        wildcard = "PDF Files (*.pdf)|*.pdf|" \
                   "All files (*.*)|*.*"

        dialog = wx.FileDialog(None, "Choose a file", os.path.expanduser("~"), "", wildcard, wx.SAVE)
        if dialog.ShowModal() == wx.ID_OK:
            path = dialog.GetPath()

        dialog.Destroy()
        logging.info("Saving from Dialog: " + path)
        shutil.move(source_file, path)

    def run_convert(self, text_recognize):
        """
        convert images to pdf file, run text recongnition (if desired),
        save to directory (pre-defined or user-defined)
        :param text_recognize: boolean on whether to apply text recognition
        :return: None
        """
        os.chdir(self.tmp_dir)
        if self.pdf_filename is not None:
            pdf_basename = os.path.basename(self.pdf_filename)
        else:
            pdf_basename = "temp.pdf"
        ps_filename = pdf_basename
        ps_filename = ps_filename.replace(".pdf", ".ps")
        c = ['convert',
             '-density %s' % self.dpi,
             ' '.join(self.pages),
             ps_filename
             ]
        self.cmd(c)
        c = ['ps2pdf',
             '-DPDFSETTINGS=/prepress',
             ps_filename,
             pdf_basename,
             ]
        self.cmd(c)
        if text_recognize:
            logging.info("running pdf sandwich for text recognition...")
            self.run_text_recognize(pdf_basename)
        if self.pdf_filename is not None:
            shutil.move(pdf_basename, self.pdf_filename)
        else:
            source_file = os.path.join(self.tmp_dir, pdf_basename)
            self.save_file(source_file)
        for filename in self.pages + [ps_filename]:
            os.remove(filename)

        # IF we did the scan, then remove the tmp dir too
        if self.args['scan'] and not self.args['--keep-tmpdir']:
            os.rmdir(self.tmp_dir)
        os.chdir(self.cwd)

    def get_options(self, argv):
        """
            Parse the command-line options and set the following object properties:

            :param argv: usually just sys.argv[1:]
            :returns: Nothing
        """
        self.args = argv

        if argv['--verbose']:
            logging.basicConfig(level=logging.INFO, format='%(message)s')
        if argv['--debug']:
            logging.basicConfig(level=logging.DEBUG, format='%(message)s')
        if self.args['pdf']:
            if self.args['<pdffile>']:
                self.pdf_filename = os.path.abspath(self.args['<pdffile>'])
                logging.info('saving to file: ' + self.pdf_filename)
            else:
                logging.info('saving file via dialog after processing...')
        self.dpi = self.args['--dpi']
        output_dir = time.strftime('%Y%m%d_%H%M%S', time.localtime())
        if argv['--tmpdir']:
            self.tmp_dir = argv['--tmpdir']
        else:
            self.tmp_dir = os.path.join('/tmp', output_dir)
        self.tmp_dir = os.path.abspath(self.tmp_dir)

        # Make the tmp dir only if we're scanning, o/w throw an error, also get device
        if argv['scan']:
            if argv['--device']:
                self.device = argv['--device'].replace('%', ' ')  # Replace % with spaces to comply with docopt
            else:
                self.device = os.environ.get('SCANBD_DEVICE')
            if os.path.exists(self.tmp_dir):
                self._error("Temporary output directory %s already exists!" % self.tmp_dir)
            else:
                os.makedirs(self.tmp_dir)
        else:
            if not os.path.exists(self.tmp_dir):
                self._error("Scan files directory %s does not exist!" % self.tmp_dir)

        # Blank checks
        self.keep_blanks = argv['--keep-blanks']
        self.blank_threshold = float(argv['--blank-threshold'])
        assert (0 <= self.blank_threshold <= 1.0)
        self.crop = argv['--crop']
        self.text_recognize = argv['--text-recognize']
        self.post_process = argv['--post-process']

    def go(self, argv):
        """
            The main entry point into ScanPdf

            #. Get the options
            #. Create the temp dir
            #. Run scanadf
        """
        # Read the command line options
        self.get_options(argv)
        start = time.time()
        logging.info("Starting: " + time.strftime(date_format))
        logging.info("Temp dir: %s" % self.tmp_dir)
        if self.args['scan']:
            self.run_scan()

        if self.args['pdf']:
            # Now, convert the files to ps
            self.pages = self.get_pages()
            logging.debug(self.pages)
            if self.args['--face-up'] == 'True':  # Default is a text value of 'True'
                self.reorder_face_up()

            start_pages = [(page, self) for page in self.pages]

            threads_to_run = multiprocessing.cpu_count()
            logging.info('Processing pages with {0:d} threads...'.format(threads_to_run))

            pool = Threadpool(threads_to_run)
            results = pool.map(ProcessPage.run, start_pages)
            pool.close()
            pool.join()

            self.pages = [page for page in results if page is not None]

            logging.debug(self.pages)
            self.run_convert(self.text_recognize)

            end = time.time()
            logging.info("End: " + time.strftime(date_format))
            logging.info("Elapsed Time (seconds): {0:.2f}".format(end-start))


def main():
    args = docopt.docopt(__doc__, version='Scan PDF %s' % __version__)
    script = ScanPdf()
    print args
    script.go(args)


if __name__ == '__main__':
    main()
